### PR TITLE
MONGOCRYPT-331 add KMIP reader and writer

### PR DESCRIPTION
**Pull request by kevinAlbs**
**Background**

Hashicorp Vault supports a subset of KMIP 1.4. I referred to the [KMIP 1.4 specification](http://docs.oasis-open.org/kmip/spec/v1.4/os/kmip-spec-v1.4-os.html) section 9.

We do not need a full implementation. The plan is to only support the create, get, and activate operations on a 96 byte SecretData managed object as the key encryption key (KEK).

Most of the credit goes to Mark. This is copied and modified from his [KMIP support prototype](https://github.com/markbenvenuto/libmongocrypt/tree/kmip).

**Scope**

- Add a `kmip_reader_t` and `kmip_writer_t` types.

### CODE REVIEW

File: kms-message/test/test_kms_request.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TEST_KMS_REQUEST_H
+#define TEST_KMS_REQUEST_H
+
+#define ASSERT_CMPSTR(_a, _b) compare_strs (__FUNCTION__, (_a), (_b))
+
+#define ASSERT(stmt)                                                        \
```
Comment (eramongodb) [inline kms-message/test/test_kms_request.h:22]: Should this macro (and others defined in this header) be given a prefix such as `TEST_KMS_` or `KMS_`?

**Review by eramongodb — CHANGES_REQUESTED** [review]

File: kms-message/src/kms_kmip_reader_writer_private.h [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+/* Note: does not descend structures. */
+bool
+kmip_reader_find (kmip_reader_t *reader,
+                  kmip_tag_type_t search_tag,
+                  kmip_item_type_t type,
+                  size_t *pos,
+                  size_t *length);
+
+kmip_reader_t *
+kmip_reader_find_and_get_struct_reader (kmip_reader_t *reader, size_t tag);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply kms-message/src/kms_kmip_reader_writer_private.h:137]: Keep, it is used in the prototype. Added a unit test.

File: kms-message/src/kms_kmip_reader_writer_private.h [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+
+bool
+kmip_reader_find_and_read_enum (kmip_reader_t *reader,
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply kms-message/src/kms_kmip_reader_writer_private.h:140]: Keep, it is used in the prototype. Added a unit test.

File: kms-message/src/kms_kmip_reader_writer_private.h [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+                                size_t tag,
+                                uint32_t *value);
+
+bool
+kmip_reader_find_and_read_bytes (kmip_reader_t *reader,
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply kms-message/src/kms_kmip_reader_writer_private.h:145]: Keep, it is used in the prototype. Added a unit test.

File: kms-message/test/test_kmip_reader_writer.c [hunk 5]
```diff
@@ -354,3 +355,156 @@ kms_kmip_reader_negative_int_test (void)
    kmip_reader_destroy (reader);
    free (data);
 }
+
+void
+kms_kmip_reader_find_test (void)
```
Comment (kevinAlbs) [inline kms-message/test/test_kmip_reader_writer.c:360]: Note: I think I should rename all of these test functions to remove the `kms_` prefix unless anyone objects. The new `kmip` types and functions do not have the `kms` prefix.

File: kms-message/src/kms_kmip_tag_type_private.h [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+   KMIP_TAG_CertificateIssuerDC = 0x42011E,
+   KMIP_TAG_CertificateIssuerDNQualifier = 0x42011F,
+   KMIP_TAG_Sensitive = 0x420120,
+   KMIP_TAG_AlwaysSensitive = 0x420121,
+   KMIP_TAG_Extractable = 0x420122,
+   KMIP_TAG_NeverExtractable = 0x420123,
+   KMIP_TAG_ReplaceExisting = 0x420124,
+} kmip_tag_type_t;
+
+#endif /* KMS_KMIP_TAG_TYPE_PRIVATE_H */
```
Comment (eramongodb) [inline kms-message/src/kms_kmip_tag_type_private.h:319]: Missing EOF newline.

File: kms-message/src/kms_endian_private.h [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+__kms_uint64_swap_slow (uint64_t v) /* IN */
+{
+   return ((v & 0x00000000000000FFULL) << 56) |
+          ((v & 0x000000000000FF00ULL) << 40) |
+          ((v & 0x0000000000FF0000ULL) << 24) |
+          ((v & 0x00000000FF000000ULL) << 8) |
+          ((v & 0x000000FF00000000ULL) >> 8) |
+          ((v & 0x0000FF0000000000ULL) >> 24) |
+          ((v & 0x00FF000000000000ULL) >> 40) |
+          ((v & 0xFF00000000000000ULL) >> 56);
```
Comment (eramongodb) [inline kms-message/src/kms_endian_private.h:161]: Note: `long long` is a new C99 feature in the mongocrypt project (supported as an extension when C99 mode is not enabled).

File: kms-message/src/kms_kmip_item_type_private.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+   KMIP_ITEM_TYPE_Structure = 0x01,
+   KMIP_ITEM_TYPE_Integer = 0x02,
+   KMIP_ITEM_TYPE_LongInteger = 0x03,
+   KMIP_ITEM_TYPE_BigInteger = 0x04,
+   KMIP_ITEM_TYPE_Enumeration = 0x05,
+   KMIP_ITEM_TYPE_Boolean = 0x06,
+   KMIP_ITEM_TYPE_TextString = 0x07,
+   KMIP_ITEM_TYPE_ByteString = 0x08,
+   KMIP_ITEM_TYPE_DateTime = 0x09,
+   KMIP_ITEM_TYPE_Interval = 0x0A,
```
Comment (eramongodb) [inline kms-message/src/kms_kmip_item_type_private.h:30]: Note: trailing comma at end of enumerator list is a new C99 feature in the mongocrypt project (supported as an extension when C99 mode is not enabled).

**Review by eramongodb — APPROVED** [review]
Some missing EOF newlines and some notes regarding new C99 feature use. Otherwise, LGTM.

File: kms-message/src/kms_kmip_tag_type_private.h [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+   KMIP_TAG_CertificateIssuerUID = 0x42011B,
+   KMIP_TAG_CertificateIssuerSerialNumber = 0x42011C,
+   KMIP_TAG_CertificateIssuerTitle = 0x42011D,
+   KMIP_TAG_CertificateIssuerDC = 0x42011E,
+   KMIP_TAG_CertificateIssuerDNQualifier = 0x42011F,
+   KMIP_TAG_Sensitive = 0x420120,
+   KMIP_TAG_AlwaysSensitive = 0x420121,
+   KMIP_TAG_Extractable = 0x420122,
+   KMIP_TAG_NeverExtractable = 0x420123,
+   KMIP_TAG_ReplaceExisting = 0x420124,
```
Comment (eramongodb) [inline kms-message/src/kms_kmip_tag_type_private.h:316]: If we want to remove trailing commas in enumerator lists, there is one more here.

**Review by markbenvenuto — APPROVED** [review]

File: kms-message/src/kms_kmip_tag_type_private.h (hunk 9 above)
Comment (kevinAlbs) [reply kms-message/src/kms_kmip_tag_type_private.h:316]: Good catch, done.

**Pull request merged**